src/test/java/core/basesyntax/RegistrationServiceImplTest.java

```diff
@@ -0,0 +1,166 @@
+
+    @BeforeEach
+    void setUp() {
+        storageDao = new StorageDaoImpl();
+        Storage.people.clear();
+    }
+
+    @Test
+    void addUser_NullLogin_ThrowsException() {
+        User user = new User();
+        user.setLogin(null);
+        user.setPassword("qwerty1234");
+        user.setAge(17);
+        assertThrows(RegistrationException.class, () -> REGISTRATION_SERVICE.register(user));
+    }
+
+    @Test
+    void exists_Login() {
+        User user = new User();
+        user.setLogin("Alice123");
+        user.setPassword("qwerty1234");
+        user.setAge(19);
+        storageDao.add(user);
+
+        User foundUser = storageDao.get("Alice123");
+        assertNotNull(foundUser);
+        assertEquals("Alice123", foundUser.getLogin());
+    }
+
+    @Test
+    void addUser_EmptyLogin_ThrowsException() {
+        User user = new User();
+        user.setLogin("");
+        user.setPassword("qwerty1234");
+        user.setAge(17);
+        assertThrows(RegistrationException.class, () -> REGISTRATION_SERVICE.register(user));
+    }
+
+    @Test
+    void addUser_NullPassword_ThrowsException() {
+        User user = new User();
+        user.setLogin("Bob1256");
+        user.setPassword(null);
+        user.setAge(25);
+        assertThrows(RegistrationException.class, () -> REGISTRATION_SERVICE.register(user));
+    }
+
+    @Test
+    void addUser_EmptyPassword_ThrowsException() {
+        User user = new User();
+        user.setLogin("Bob1256");
+        user.setPassword("");
+        user.setAge(25);
+        assertThrows(RegistrationException.class, () -> REGISTRATION_SERVICE.register(user));
+    }
+
+    @Test
+    void addUser_LoginLessThanMinLength_ThrowsException() {
+        User user = new User();
+        user.setLogin("Bob");
+        user.setPassword("qwerty1234");
+        user.setAge(25);
+        assertThrows(RegistrationException.class, () -> REGISTRATION_SERVICE.register(user));
+    }
+
+    @Test
+    void addUser_PasswordLessThanMinLength_ThrowsException() {
+        User user = new User();
+        user.setLogin("Bob1256");
+        user.setPassword("qwert");
+        user.setAge(25);
+        assertThrows(RegistrationException.class, () -> REGISTRATION_SERVICE.register(user));
+    }
+
+    @Test
+    void addUser_Underage_ThrowsException() {
+        User user = new User();
+        user.setLogin("Alice");
+        user.setPassword("qwerty1234");
+        user.setAge(17);
+        assertThrows(RegistrationException.class, () -> REGISTRATION_SERVICE.register(user));
+    }
+
+    @Test
+    void addUser_WithValidDetails_Success() {
+        User user = new User();
+        user.setPassword("qwerty123");
+        user.setLogin("Alex1976");
+        user.setAge(27);
+        User registeredUser = REGISTRATION_SERVICE.register(user);
+        assertEquals(user, registeredUser);
+    }
+
+    @Test
+    void addUser_InvalidAge_ThrowsException() {
```

Comment:
The test cases should also cover scenarios where the user's age is exactly 18, which is an edge case for the minimum age requirement.